zipkin/src/main/java/zipkin2/internal/ReadBuffer.java

```diff
@@ -0,0 +1,384 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package zipkin2.internal;
+
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+import static zipkin2.internal.HexCodec.HEX_DIGITS;
+import static zipkin2.internal.JsonCodec.UTF_8;
+
+/** Read operations do bounds checks, as typically more errors occur reading than writing. */
+public abstract class ReadBuffer extends InputStream {
+
+  /** Do not use the buffer passed here after, as it may be manipulated directly. */
+  public static ReadBuffer wrapUnsafe(ByteBuffer buffer) {
+    if (buffer.hasArray()) {
+      int offset = buffer.arrayOffset() + buffer.position();
+      int limit = offset + buffer.remaining();
+      return wrap(buffer.array(), offset, limit);
+    }
+    return new ReadBuffer.Buff(buffer);
+  }
+
+  public static ReadBuffer wrap(byte[] bytes) {
+    return wrap(bytes, 0, bytes.length);
+  }
+
+  public static ReadBuffer wrap(byte[] bytes, int pos, int limit) {
+    return new ReadBuffer.Array(bytes, pos, limit);
+  }
+
+  static final class Buff extends ReadBuffer {
+    final ByteBuffer buf; // visible for testing
+
+    Buff(ByteBuffer buf) {
+      this.buf = buf;
+    }
+
+    @Override final byte readByteUnsafe() {
+      return buf.get();
+    }
+
+    @Override final byte[] readBytes(int length) {
+      require(length);
+      byte[] copy = new byte[length];
+      buf.get(copy);
+      return copy;
+    }
+
+    @Override boolean tryReadAscii(char[] destination, int length) {
+      buf.mark();
+      for (int i = 0; i < length; i++) {
+        byte b = buf.get();
+        if ((b & 0x80) != 0) {
+          buf.reset();
+          return false;  // Not 7-bit ASCII character
+        }
+        destination[i] = (char) b;
+      }
+      return true;
+    }
+
+    @Override final String doReadUtf8(int length) {
+      return new String(readBytes(length), UTF_8);
+    }
+
+    @Override short readShort() {
+      require(2);
+      return buf.getShort();
+    }
+
+    @Override int readInt() {
+      require(4);
+      return buf.getInt();
+    }
+
+    @Override long readLong() {// TODO: test me with flipped order
+      require(8);
+      return getLong(ByteOrder.BIG_ENDIAN);
+    }
+
+    @Override long readLongLe() {
+      require(8);
+      return getLong(ByteOrder.LITTLE_ENDIAN);
+    }
+
+    long getLong(ByteOrder wanted) {
+      if (buf.order() == wanted) return buf.getLong();
+      return (readByteUnsafe() & 0xffL)
```

Comment:
I think this always returns little endian, even if someone called readLong on a little endian buffer. Though maybe the TODO already expects the test to fail? ;)

Reply:
hah sorry this was sloppy of me. will fix.. I still owe some unit tests on this one

Reply:
tests done